cpp/src/arrow/compute/exec/bloom_filter_test.cc

```diff
@@ -0,0 +1,633 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <gmock/gmock-matchers.h>
+
+#include <algorithm>
+#include <chrono>
+#include <condition_variable>
+#include <set>
+#include <thread>
+#include "arrow/compute/exec/bloom_filter.h"
+#include "arrow/compute/exec/key_hash.h"
+#include "arrow/compute/exec/test_util.h"
+#include "arrow/compute/exec/util.h"
+#include "arrow/util/bitmap_ops.h"
+#include "arrow/util/cpu_info.h"
+
+namespace arrow {
+namespace compute {
+
+TEST(BloomFilter, Masks) {
+  BloomFilterMasks masks;
+  std::vector<int> sum_num_masks_with_same_n_bits(BloomFilterMasks::kMaxBitsSet + 1);
+  std::vector<int> num_masks_with_same_n_bits(BloomFilterMasks::kMaxBitsSet + 1);
+
+  for (bool with_rotation : {false, true}) {
+    printf("With bit rotation: %s\n", with_rotation ? "ON" : "OFF");
+
+    for (int i = 0; i <= BloomFilterMasks::kMaxBitsSet; ++i) {
+      sum_num_masks_with_same_n_bits[i] = 0;
+    }
+
+    for (int imask = 0; imask < BloomFilterMasks::kNumMasks; ++imask) {
+      uint64_t mask = masks.mask(imask);
+      // Verify that the number of bits set is in the required range
+      //
+      ARROW_DCHECK(ARROW_POPCOUNT64(mask) >= BloomFilterMasks::kMinBitsSet &&
+                   ARROW_POPCOUNT64(mask) <= BloomFilterMasks::kMaxBitsSet);
```

Comment:
Why is the second form better than the first one?

Reply:
Pretty minor differences:

`DCHECK` only runs on debug mode where `ASSERT` checks on both debug and release mode (our CI runs the tests in 
both).  `DCHECK` will also trigger an abort which will prevent any other test cases from running.

It would be fine as a single assert:
```
ASSERT_TRUE(ARROW_POPCOUNT64(mask) >= BloomFilterMasks::kMinBitsSet &&
                   ARROW_POPCOUNT64(mask) <= BloomFilterMasks::kMaxBitsSet);
```
...but making it two asserts would help if it ever failed to narrow down what actually failed (i.e. was the result too large or too small?)